src/zkevm_specs/state.py

```diff
@@ -0,0 +1,544 @@
+
+
+@is_circuit_code
+def check_call_context(row: Row, row_prev: Row):
+    get_call_id = lambda row: row.keys[1]
+    get_field_tag = lambda row: row.keys[3]
+
+    # 0. Unused keys are 0
+    assert row.keys[2] == 0
+    assert row.keys[4] == 0
+
+    # TODO: Missing constraints
+
+
+@is_circuit_code
+def check_account(row: Row, row_prev: Row):
+    get_addr = lambda row: row.keys[2]
+    get_field_tag = lambda row: row.keys[3]
+
+    # 0. Unused keys are 0
+    assert row.keys[1] == 0
+    assert row.keys[4] == 0
+
+    # 1. First access for a set of all keys
+    #
+    # We add an extra write to setup the value of the previous block, with rwc=0.
+    #
+    # When the set of all keys changes (first access of storage (address, AccountFieldTag))
+    # - It must be a WRITE
+    if not all_keys_eq(row, row_prev):
+        assert row.is_write == 1 and row.rw_counter == 0
+
+    # 2. nonce
+    if get_field_tag(row) == AccountFieldTag.Nonce:
+        get_nonce = lambda row: row.value
+        # When all the keys are equal in two consecutive a rows:
+        # - if WRITE, Nonce increases by 1
+        if all_keys_eq(row, row_prev) and row.is_write == 1:
+            assert get_nonce(row) - get_nonce(row_prev) == 1
+    elif get_field_tag(row) == AccountFieldTag.Balance:
+        get_balance = lambda row: row.value
+        pass
+    elif get_field_tag(row) == AccountFieldTag.CodeHash:
+        get_code_hash = lambda row: row.value
+        pass
+
+
+@is_circuit_code
+def check_tx_refund(row: Row, row_prev: Row):
+    get_tx_id = lambda row: row.keys[1]
+
+    # 0. Unused keys are 0
+    assert row.keys[2] == 0
+    assert row.keys[3] == 0
+    assert row.keys[4] == 0
+
+    # TODO: Missing constraints
+
+
+@is_circuit_code
+def check_tx_access_list_account(row: Row, row_prev: Row):
+    get_tx_id = lambda row: row.keys[1]
+    get_addr = lambda row: row.keys[2]
+
+    # 0. Unused keys are 0
+    assert row.keys[3] == 0
+    assert row.keys[4] == 0
+
+    # TODO: Missing constraints
+
+
+@is_circuit_code
+def check_tx_access_list_account_storage(row: Row, row_prev: Row):
+    get_tx_id = lambda row: row.keys[1]
+    get_addr = lambda row: row.keys[2]
+    get_storage_key = lambda row: row.keys[4]
+
+    # 0. Unused keys are 0
+    assert row.keys[3] == 0
+
+    # TODO: Missing constraints
+
+
+@is_circuit_code
+def check_account_destructed(row: Row, row_prev: Row):
+    get_addr = lambda row: row.keys[2]
+
+    # 0. Unused keys are 0
+    assert row.keys[1] == 0
+    assert row.keys[3] == 0
+    assert row.keys[4] == 0
+
+    # TODO: Missing constraints
+
+
+@is_circuit_code
+def check_state_row(row: Row, row_prev: Row, randomness: FQ):
+    #
+    # Constraints that affect all rows, no matter which Tag they use
+    #
+
+    # 0. Tags transition
+    #
+    # Tag monotonically increases for all values except for Start
+    if row.tag() != Tag.Start:
+        tag_diff = row.tag() - row_prev.tag()
+        assert_in_range(tag_diff, 0, MAX_KEY_DIFF)
```

Comment:
Can we be sure that all blocks will use all tags always?
If that's not the case, then there could be an increment of more than 1.  Assuming that was the case, I just went to use `MAX_KEY_DIFF` to avoid introducing a new range lookup table.  I'm not familiar with the cost of A vs B:
- A. 1 big fixed lookup table with 2 lookups happening
- B. 1 big fixed lookup table with 1 lookup + 1 small fixed lookup table with 1 lookup

Reply:
I see, it makes sense to me now.

Since we already have different range lookup, we would somehow have tag for fixed table to reuse the fixed columns. If we do so, then cost of A vs B is same.

Reply:
I have added constraints for each key to be in the expected range (including key0 = tag) in https://github.com/appliedzkp/zkevm-specs/compare/b145f36feaab9db33f41c4cd14251dc961b7609d..0258fe56111ed78f653b02a51d08093ec05958d0